### PR TITLE
[Fabric] feat: Adds fabric iOS minimal version of screens

### PR DESCRIPTION
**Pull request by Ubax**
## Description

Adds fabric iOS minimal version of screens

## Checklist

- [ ] Ensured that CI passes

### CODE REVIEW

**Comment by robertjpayne** [comment]
@Ubax tried this out as I'm attempting to use Fabric… just so you're aware the "pop" animation on a `ScreenStack` removes the screen before performing the native animation. If you toggle slow animations on iOS Simulator you can see the effects pretty easily.

**Comment by Ubax** [comment]
@robertjpayne thanks for looking into this PR. 

We are aware about this problem and currently thinking about best solution for that. Did you find any more problems?

**Comment by robertjpayne** [comment]
Hey @Ubax I didn't test it much more, was actually not possible to get it working outside of Fabric example. trying to `import {...} from 'react-native-screens/fabric' just gives me errors. I'm more a iOS/Swift/C++ developer than JavaScript though.

Because of that I checked out the example project and found the bug. Seems really tricky to fix since react is ditching the view, might need to snapshot it during the transition out? Not sure how that would work on Android though.

File: package.json [hunk 1]
```diff
@@ -126,5 +126,14 @@
   "eslintIgnore": [
     "node_modules/",
     "lib/"
-  ]
-}
+  ],
+  "codegenConfig": {
+    "libraries": [
+      {
+        "name": "ScreensSpec",
```
Comment (tomekzaw) [inline package.json:133]:
```suggestion
        "name": "rnscreens",
```

File: FabricExample/ios/Podfile [hunk 2]
```diff
@@ -1,20 +1,21 @@
+install! 'cocoapods', :deterministic_uuids => false
+
 require_relative '../node_modules/react-native/scripts/react_native_pods'
 require_relative '../node_modules/@react-native-community/cli-platform-ios/native_modules'
 
 platform :ios, '11.0'
 
+fabric_enabled = ENV['USE_FABRIC']
+use_codegen_discovery= ENV['USE_CODEGEN_DISCOVERY']
```
[the remark this answers is not in the document]
Comment (Ubax) [reply FabricExample/ios/Podfile:9]: I will just use the Podfile from RN 68

File: common/cpp/rnscreens/RNSScreenComponentDescriptor.h [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+        std::dynamic_pointer_cast<YogaLayoutableShadowNode>(screenShadowNode));
+    auto layoutableShadowNode =
+        std::static_pointer_cast<YogaLayoutableShadowNode>(screenShadowNode);
+
+    auto state =
+        std::static_pointer_cast<const RNSScreenShadowNode::ConcreteState>(
+            shadowNode->getState());
+    auto stateData = state->getData();
+
+    // TODO: Test if this work on iOS
```
[the remark this answers is not in the document]
Comment (Ubax) [reply common/cpp/rnscreens/RNSScreenComponentDescriptor.h:36]: This TODO is actually done 😄

File: common/cpp/rnscreens/RNSScreenShadowNode.cpp [hunk 4]
```diff
@@ -0,0 +1,11 @@
+#include "RNSScreenShadowNode.h"
```
[the remark this answers is not in the document]
Comment (Ubax) [reply common/cpp/rnscreens/RNSScreenShadowNode.cpp:1]: Unfortunately we cannot. 
We are using `interfaceOnly: true,` flag so it is not generated

File: FabricExample/babel.config.js [hunk 5]
```diff
@@ -1,3 +1,13 @@
 module.exports = {
-  presets: ['module:metro-react-native-babel-preset'],
+  presets: ['babel-preset-expo'],
```
[the remark this answers is not in the document]
Comment (Ubax) [reply FabricExample/babel.config.js:2]: It was copied from Example application

[the remark this answers is not in the document]
Comment (Ubax) [reply FabricExample/babel.config.js:2]: I'll test it with `module:metro-react-native-babel-preset` and if works change to it

File: ios/RNSScreenStackComponentView.mm [hunk 6]
```diff
@@ -0,0 +1,292 @@
...
+- (void)maybeAddToParentAndUpdateContainer
+{
+  BOOL wasScreenMounted = _controller.parentViewController != nil;
+  BOOL isScreenReadyForShowing = self.window;
+  if (!isScreenReadyForShowing && !wasScreenMounted) {
+    return;
+  }
+  [self updateContainer];
+  if (!wasScreenMounted) {
+    // when stack hasn't been added to parent VC yet we do two things:
```
[the remark this answers is not in the document]
Comment (Ubax) [reply ios/RNSScreenStackComponentView.mm:135]: The same comment is in `RNSScreenStack.m` and I would like to keep these files as much in sync as possible.
Next step would be to write common logic in one file and then we can change it.

File: ios/RNSScreenStackComponentView.mm [hunk 7]
```diff
@@ -0,0 +1,292 @@
...
+- (void)updateContainer
+{
+  NSMutableArray<UIViewController *> *pushControllers = [NSMutableArray new];
+  for (RNSScreenComponentView *screen in _reactSubviews) {
+    if (screen.controller != nil) {
+      if (pushControllers.count == 0) {
+        // first screen on the list needs to be places as "push controller"
+        [pushControllers addObject:screen.controller];
+      } else {
+        [pushControllers addObject:screen.controller];
```
[the remark this answers is not in the document]
Comment (Ubax) [reply ios/RNSScreenStackComponentView.mm:239]: It similar to the comment above. I would like to keep the files as in sync as possible

File: ios/RNSScreenStackComponentView.mm [hunk 8]
```diff
@@ -0,0 +1,292 @@
...
+{
+  return YES;
+}
+
+#pragma mark - RCTComponentViewProtocol
+
+- (void)prepareForRecycle
+{
+  [super prepareForRecycle];
+  _reactSubviews = [NSMutableArray new];
```
[the remark this answers is not in the document]
Comment (Ubax) [reply ios/RNSScreenStackComponentView.mm:279]: We cannot deallocate it here. 
We also don't need to create it here as it has same lifecycle as the whole ScreenStack itself

File: ios/RNSScreenStackHeaderConfigComponentView.h [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+
+@interface RNSScreenStackHeaderConfigComponentView : RCTViewComponentView
+
+@property (nonatomic, weak) RNSScreenComponentView *screenView;
+
+@property (nonatomic) BOOL show;
+@property (nonatomic) BOOL translucent;
+@property (nonatomic) NSMutableArray<RNSScreenStackHeaderSubviewComponentView *> *reactSubviews;
+
++ (void)willShowViewController:(UIViewController *)vc
```
[the remark this answers is not in the document]
Comment (Ubax) [reply ios/RNSScreenStackHeaderConfigComponentView.h:16]: It's again to keep in sync with old architecture

File: ios/RNSScreenStackHeaderConfigComponentView.mm [hunk 10]
```diff
@@ -0,0 +1,234 @@
...
+    vc.edgesForExtendedLayout = UIRectEdgeAll;
+  }
+
+  [navctr setNavigationBarHidden:shouldHide animated:animated];
+
+  if (shouldHide) {
+    return;
+  }
+
+  navitem.title = @"test title";
```
[the remark this answers is not in the document]
Comment (Ubax) [reply ios/RNSScreenStackHeaderConfigComponentView.mm:117]: Nope

File: common/cpp/rnscreens/RNSScreenShadowNode.cpp [hunk 11]
```diff
@@ -0,0 +1,11 @@
...
+
+#include <react/renderer/core/LayoutContext.h>
```
[the remark this answers is not in the document]
Comment (Ubax) [reply common/cpp/rnscreens/RNSScreenShadowNode.cpp]: No it's not
Thanks

File: ios/RNSScreenComponentView.mm [hunk 12]
```diff
@@ -0,0 +1,148 @@
+#import "RNSScreenComponentView.h"
+#import "RNSScreenStackHeaderConfigComponentView.h"
+
+#import <React/RCTConversions.h>
+
+#import <react/renderer/components/ScreensSpec/EventEmitters.h>
+#import <react/renderer/components/ScreensSpec/Props.h>
+#import <react/renderer/components/ScreensSpec/RCTComponentViewHelpers.h>
+#import "../common/cpp/rnscreens/RNSScreenComponentDescriptor.h"
```
[the remark this answers is not in the document]
Comment (Ubax) [reply ios/RNSScreenComponentView.mm]: We cannot change `../common/cpp/rnscreens/RNSScreenComponentDescriptor.h` to `react/renderer/components/rnscreens/ComponentDescriptors.h`. This is custom component descriptor.

[the remark this answers is not in the document]
Comment (Ubax) [reply ios/RNSScreenComponentView.mm]: I changed the relative path to the absolute one though

File: ios/RNSScreenComponentView.mm [hunk 13]
```diff
@@ -0,0 +1,148 @@
...
+  }
+}
+
+#pragma mark - RCTComponentViewProtocol
+
+- (void)prepareForRecycle
+{
+  [super prepareForRecycle];
+  // TODO: Make sure that there is no edge case when this should be uncommented
+  // _controller=nil;
```
[the remark this answers is not in the document]
Comment (Ubax) [reply ios/RNSScreenComponentView.mm:122]: It still remains TODO. Not all the logic is rewritten and ready yet

File: ios/RNSScreenControler.mm [hunk 14]
```diff
@@ -0,0 +1,56 @@
...
+
+- (instancetype)initWithView:(UIView *)view
+{
+  if (self = [super init]) {
+    self.view = view;
+  }
+  return self;
+}
+
+// TODO: Find out why this is executed when screen is going out
```
[the remark this answers is not in the document]
Comment (Ubax) [reply ios/RNSScreenControler.mm:15]: It still remains TODO. Not all the logic is rewritten and ready yet

**Comment by Ubax** [comment]
@tomekzaw I answered some of your comments
Thank you very much for the review

**Pull request closed without merging**